src/erofs/format.rs

```diff
@@ -0,0 +1,279 @@
+
+impl Default for FormatField {
+    fn default() -> Self {
+        Self(0xffff.into())
+    }
+}
+
+impl std::fmt::Debug for FormatField {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{} = {:?} | {:?}",
+            self.0.get(),
+            InodeLayout::from(*self),
+            DataLayout::try_from(*self)
+        )
+    }
+}
+
+const INODE_LAYOUT_MASK: u16 = 0b00000001;
+const INODE_LAYOUT_COMPACT: u16 = 0;
+const INODE_LAYOUT_EXTENDED: u16 = 1;
+
+#[derive(Debug)]
+#[repr(u16)]
+pub enum InodeLayout {
+    Compact = INODE_LAYOUT_COMPACT,
+    Extended = INODE_LAYOUT_EXTENDED,
+}
+
+impl From<FormatField> for InodeLayout {
+    fn from(value: FormatField) -> Self {
+        match value.0.get() & INODE_LAYOUT_MASK {
+            INODE_LAYOUT_COMPACT => InodeLayout::Compact,
+            INODE_LAYOUT_EXTENDED => InodeLayout::Extended,
+            _ => unreachable!(),
+        }
+    }
+}
+
+const INODE_DATALAYOUT_MASK: u16 = 0b00001110;
+const INODE_DATALAYOUT_FLAT_PLAIN: u16 = 0;
+const INODE_DATALAYOUT_FLAT_INLINE: u16 = 4;
+const INODE_DATALAYOUT_CHUNK_BASED: u16 = 8;
+
+#[derive(Debug)]
+#[repr(u16)]
+pub enum DataLayout {
+    FlatPlain = 0,
+    FlatInline = 4,
+    ChunkBased = 8,
+}
+
+impl TryFrom<FormatField> for DataLayout {
+    type Error = FormatError;
+
+    fn try_from(value: FormatField) -> Result<Self, FormatError> {
+        match value.0.get() & INODE_DATALAYOUT_MASK {
+            INODE_DATALAYOUT_FLAT_PLAIN => Ok(DataLayout::FlatPlain),
+            INODE_DATALAYOUT_FLAT_INLINE => Ok(DataLayout::FlatInline),
+            INODE_DATALAYOUT_CHUNK_BASED => Ok(DataLayout::ChunkBased),
+            _ => Err(FormatError::InvalidDataLayout),
+        }
+    }
+}
+
+impl From<(InodeLayout, DataLayout)> for FormatField {
+    fn from(value: (InodeLayout, DataLayout)) -> FormatField {
+        FormatField(
+            (match value.0 {
+                InodeLayout::Compact => INODE_LAYOUT_COMPACT,
+                InodeLayout::Extended => INODE_LAYOUT_EXTENDED,
+            } | match value.1 {
+                DataLayout::FlatPlain => INODE_DATALAYOUT_FLAT_PLAIN,
+                DataLayout::FlatInline => INODE_DATALAYOUT_FLAT_INLINE,
+                DataLayout::ChunkBased => INODE_DATALAYOUT_CHUNK_BASED,
+            })
+            .into(),
+        )
+    }
+}
+
+/* Extended attributes */
+pub const XATTR_FILTER_SEED: u32 = 0x25BBE08F;
+
+#[derive(Debug, Immutable, IntoBytes, KnownLayout, TryFromBytes)]
+#[repr(C)]
+pub struct XAttrHeader {
+    pub name_len: u8,
+    pub name_index: u8,
+    pub value_size: U16,
+}
+
+pub const XATTR_PREFIXES: [&[u8]; 7] = [
+    b"",
+    b"user.",
+    b"system.posix_acl_access",
+    b"system.posix_acl_default",
+    b"trusted.",
+    b"lustre.",
+    b"security.",
+];
+
+/* Directories */
+
+#[derive(Clone, Copy, Debug, Default, Immutable, IntoBytes, TryFromBytes)]
+#[repr(u8)]
+pub enum FileType {
```

Comment:
Why not use https://docs.rs/rustix/latest/rustix/fs/enum.FileType.html ?

Ohh, is it for the `IntoBytes`?

Reply:
Yes.  This is a mapping of integer values used by the erofs format.